V2.17.0

cbouss opened the pull request:
tensorflow 2.17.0

**Destination channel:** main

### Links

- https://github.com/tensorflow/tensorflow/tree/v2.17.0
- Relevant dependency PRs:
  - https://github.com/AnacondaRecipes/python-flatbuffers-feedstock/pull/1
  - https://github.com/AnacondaRecipes/flatbuffers-feedstock/pull/4
  - https://github.com/AnacondaRecipes/tensorboard-feedstock/pull/1
  - https://github.com/AnacondaRecipes/ml_dtypes-feedstock/pull/3
  - https://github.com/AnacondaRecipes/libml_dtypes-headers-feedstock/pull/1
  - https://github.com/AnacondaRecipes/rsync-feedstock/pull/2 (for complex file copies in build.sh) 
  - https://github.com/AnacondaRecipes/snappy-feedstock/pull/5 (with rtti enabled, for linux builds)
  - https://github.com/AnacondaRecipes/keras-feedstock/pull/4 (keras >=3.2)
  - https://github.com/AnacondaRecipes/bazel-feedstock/pull/19

### Explanation of changes:
- Moving back recipe from tensorflow_recipes to here.
- Sync with CF.
- abs.yaml configured to not build cuda on CI
- Building with XNN option xnn_enable_avx512fp16=false as GCC 11.8 does not support avx512fp16
- Building with jpeg, not libturbo_jpeg
- Updated keras required.
- Snappy with rtti required on linux.
- Updated bazel to 6.5
- Updated bazel toolchain scripts based on bazel-toolchain 0.2.0 (See https://github.com/AnacondaRecipes/bazel-toolchain-feedstock/pull/2)
- Make python handling less hermetic - packages are still downloaded from pypi at build time, but our packages are used at run time. This results in a number of overdepending warnings.

### Platforms working

- [x] linux-64
- [x] linux-64 cuda
- [x] linux-aarch64
- [x] osx-64
- [x] osx-arm64
- [ ] win-64 - to be done on subsequent PR.

cbouss committed:
adapt recipe

cbouss committed:
test

cbouss committed:
test

cbouss committed:
test

cbouss committed:
test

cbouss committed:
snappy 1.2.1

cbouss committed:
bazel 6.5.0

danpetry committed:
add deps for building wheel for estimator; only build on linux-64 to save time; build all pythons

danpetry committed:
Enable linux-aarch64, include dep that bazel's trying to fetch on aarch64, whitelist a DSO for the reason given

cbouss committed:
update toolchain

cbouss committed:
tf-base in build

cbouss committed:
loosen hermetic python

cbouss committed:
set keras tensorflow explicitly

cbouss committed:
higher build number for cuda variants

cbouss committed:
osx-64 test configuration

cbouss committed:
add libtensorflow_framework.2.dylib to missing dso

cbouss committed:
adjust for win

danpetry committed:
Some fixes for the cuda variant

danpetry commented:
The cuda packages are built, can provide them when ready to upload rest of stuff.

JeanChristopheMorinPerso commented:
Would it be possible to split up the conda-forge sync from the rest of the changes? It would make the whole review a lot easier.

cbouss commented:
> Would it be possible to split up the conda-forge sync from the rest of the changes? It would make the whole review a lot easier.

Sure, here is a PR with the sync changes only: https://github.com/AnacondaRecipes/tensorflow-feedstock/pull/2

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
Traceback (most recent call last):
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/percy/parser/recipe_parser.py", line 117, in _parse_yaml
    output = cast(JsonType, yaml.safe_load(s))
                            ^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/yaml/__init__.py", line 125, in safe_load
    return load(stream, SafeLoader)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/yaml/__init__.py", line 81, in load
    return loader.get_single_data()
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/yaml/constructor.py", line 49, in get_single_data
    node = self.get_single_node()
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/yaml/composer.py", line 35, in get_single_node
    if not self.check_event(StreamEndEvent):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/yaml/parser.py", line 98, in check_event
    self.current_event = self.state()
                         ^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/yaml/parser.py", line 142, in parse_implicit_document_start
    if not self.check_token(DirectiveToken, DocumentStartToken,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/yaml/scanner.py", line 116, in check_token
    self.fetch_more_tokens()
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/yaml/scanner.py", line 258, in fetch_more_tokens
    raise ScannerError("while scanning for the next token", None,
yaml.scanner.ScannerError: while scanning for the next token
found character '%' that cannot start any token
  in "<unicode string>", line 1, column 2:
    {% set tensorflow_cpu_gpu = "tens ...
     ^

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/anaconda_linter/run.py", line 208, in main
    return_code, report = execute_linter(
                          ^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/anaconda_linter/run.py", line 180, in execute_linter
    result = linter.lint(recipes, subdir, variant_config_files, exclusive_config_files, fix_flag)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/anaconda_linter/lint/__init__.py", line 728, in lint
    msgs = self.lint_file(
           ^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/anaconda_linter/lint/__init__.py", line 834, in lint_file
    self.lint_recipe(
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/anaconda_linter/lint/__init__.py", line 894, in lint_recipe
    res = self.check_instances[check].run(recipe, fix)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/anaconda_linter/lint/__init__.py", line 294, in run
    self.check_recipe(recipe)
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/anaconda_linter/lint/check_syntax.py", line 35, in check_recipe
    ro_parser: Final[RecipeParser] = recipe.get_read_only_parser()
                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/percy/render/recipe.py", line 675, in get_read_only_parser
    return RecipeParser("\n".join(self.meta_yaml))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/percy/parser/recipe_parser.py", line 328, in __init__
    new_node = RecipeParser._parse_line_node(clean_line)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/percy/parser/recipe_parser.py", line 147, in _parse_line_node
    output = RecipeParser._parse_yaml(s)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/percy/parser/recipe_parser.py", line 126, in _parse_yaml
    cast(JsonType, yaml.safe_load(s)), lambda d: substitute_markers(d, sub_list)
                   ^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/yaml/__init__.py", line 125, in safe_load
    return load(stream, SafeLoader)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/yaml/__init__.py", line 81, in load
    return loader.get_single_data()
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/yaml/constructor.py", line 49, in get_single_data
    node = self.get_single_node()
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/yaml/composer.py", line 35, in get_single_node
    if not self.check_event(StreamEndEvent):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/yaml/parser.py", line 98, in check_event
    self.current_event = self.state()
                         ^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/yaml/parser.py", line 142, in parse_implicit_document_start
    if not self.check_token(DirectiveToken, DocumentStartToken,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/yaml/scanner.py", line 116, in check_token
    self.fetch_more_tokens()
  File "/opt/conda/envs/linter/lib/python3.12/site-packages/yaml/scanner.py", line 258, in fetch_more_tokens
    raise ScannerError("while scanning for the next token", None,
yaml.scanner.ScannerError: while scanning for the next token
found character '%' that cannot start any token
  in "<unicode string>", line 1, column 2:
    {% set tensorflow_cpu_gpu = "tens ...
     ^

ERROR conda.cli.main_run:execute(125): `conda run conda-lint /tmp/abs_64h1h801od/clone` failed. (See above for error)
</details>

JeanChristopheMorinPerso commented on recipe/build_pkg.sh:
```diff
@@ -3,7 +3,7 @@
 
 # install the whl making sure to use host pip/python if cross-compiling
-${PYTHON} -m pip install --no-deps $SRC_DIR/tensorflow_pkg/*.whl
+${PYTHON} -m pip install --no-deps --no-build-isolation $SRC_DIR/tensorflow_pkg/*.whl
```
`--no-build-isolation` is technically redundant here since wheels are not built.

cbouss replied:
Agreed but that made the linter happy and is not harmful.

danpetry commented on recipe/build.sh:
```diff
@@ -2,6 +2,16 @@
+
 if [[ "$CI" == "github_actions" ]]; then
   export CPU_COUNT=4
 fi
```
libjpeg_turbo is mentioned in TF_SYSTEM_LIBS, does that need to be changed to jpeg?

cbouss replied:
The tensorflow build files expect this target to be named libjpeg_turbo. The code uses the libjpeg API though (jpeglib.h), which both libjpeg and libjpeg_turbo support.

danpetry replied:
right. ok thanks

danpetry commented on recipe/meta.yaml:
```diff
@@ -103,40 +108,40 @@ requirements:
+    - openssl {{ openssl }}
+    - pybind11 2
+    - sqlite {{ sqlite }}
+    - snappy 1.2.1
```
```suggestion
    # snappy needs rtti enabled, this version of ours has it
    - snappy 1.2.1
```

danpetry commented on recipe/meta.yaml:
```diff
@@ -191,24 +201,24 @@ outputs:
+        - libprotobuf-python-headers {{ libprotobuf }}
+        - openssl {{ openssl }}
+        - pybind11 2
+        - snappy 1.2.1
```
```suggestion
        # snappy needs rtti enabled, this version of ours has it
        - snappy 1.2.1
```

danpetry commented on recipe/meta.yaml:
```diff
@@ -364,26 +380,26 @@ outputs:
+        - libpng {{ libpng }}
+        - libprotobuf {{ libprotobuf }}
+        - openssl {{ openssl }}
+        - snappy 1.2.1
```
```suggestion
        # snappy needs rtti enabled, this version of ours has it
        - snappy 1.2.1
```

danpetry commented on recipe/meta.yaml:
```diff
@@ -410,28 +426,27 @@ outputs:
+        - libpng {{ libpng }}
+        - libprotobuf {{ libprotobuf }}
+        - openssl {{ openssl }}
+        - snappy 1.2.1
```
```suggestion
        # snappy needs rtti enabled, this version of ours has it
        - snappy 1.2.1
```

danpetry commented on recipe/custom_toolchain/BUILD:
```diff
@@ -7,12 +7,15 @@ filegroup(
 
 filegroup(
     name = "cc_wrapper",
-    srcs = ["cc_wrapper.sh"],
```
with this bazel toolchain stuff, it looks like this is the updated (v0.2.0) toolchain files, plus modifications that were done specifically for tensorflow. i.e. you diffed the custom_toolchain folder in conda-forge's tensorflow-feedstock with the version of bazel-feedstock it was copied from; then updated the custom_toolchain here to our latest; then re-applied these diffs? Visually, the only thing I caught extra here is the patchelf bit..?

cbouss replied:
Yes, this matches what I remember.

danpetry replied:
ok, cool. It could be nice to document this approach for whoever does the next tensorflow update. Even better, some way of using the normal toolchain files and adding in the diff somehow (maybe using sed in the build.sh script?) would be good

cbouss replied:
The way forward should be to update the normal toolchain and use it.

danpetry replied:
would that mean patchelf needs to be included in the deps for any recipes that use it? If not, is there scope for updating the normal toolchain and using it before merging this PR?

danpetry reviewed (commented):
Partial review, reviewed the items in the "explanation of changes" list, not including the one about hermetic python.

General comments:
- It would be nice to move the other tensorflow repos to the deprecated repo
- This isn't a fork of conda-forge's recipe, which we can't change now, but maybe we could add something in the README to that effect, i.e. something to tell whoever works on this next that they should pull conda-forge's changes with changes made in this PR made again on top of that.

cbouss commented on recipe/README.md:
```diff
@@ -1,3 +1,10 @@
+
+We are moving back to using this feedstock as of tensorflow 2.17.0.
+
+
```
```suggestion
This isn't a fork of conda-forge's feedstock. Whoever works on this and wishes to bring changes from the conda-forge feedstock should pull conda-forge's changes and reapply changes made in this PR on top of that.
```

cbouss committed:
Apply suggestions from code review

Address comments

Co-authored-by: danpetry <dpetry@anaconda.com>

cbouss commented:
> * It would be nice to move the other tensorflow repos to the deprecated repo

Will do once this is merged in.

> * This isn't a fork of conda-forge's recipe, which we can't change now, but maybe we could add something in the README to that effect, i.e. something to tell whoever works on this next that they should pull conda-forge's changes with changes made in this PR made again on top of that.

Added comment to README.

danpetry commented on recipe/meta.yaml:
```diff
@@ -42,49 +46,50 @@ source:
+  number: {{ build }}
+  skip: true  # [win or s390x]
   skip: true  # [python_impl == 'pypy']
   skip: true  # [py<39]
```
might want to make a note that it's untested with python 3.13 in case someone tries to rebuild this for that https://www.tensorflow.org/install/source#tested_build_configurations

cbouss replied:
As I understand, their first 3.13 version will be tensorflow 2.19. I don't think this would build with 3.13.

danpetry replied:
ok, thanks, then a skip py>312 perhaps?

danpetry commented on recipe/patches/0032-loosen-hermetic-python.patch:
```diff
@@ -0,0 +1,185 @@
+Even if packages are installed, packages are downloaded from pypi at 
+build time.
+On linux-aarch64, a number of requested packages are not available.
+This removes some declared dependencies allowing the build to proceed.
```
I've been looking into this hermetic python business and haven't understood what's going on properly (although I understand the concept in general) but was wondering if you knew the following?
- if "hermetic python" is switched on, why doesn't that mean that the tf build system will download its own python and python dependencies rather than using ours? Do we want to switch hermetic python off?
- how do the bazel files provided in this feedstock interact with the bazel files in the tf build system? Are they copied over similar files in the tree or something like that?
Basically I'm curious about why removing these pypi downloads is inconsequential. I.e. how the tf build system knows to get our dependencies anyway.

cbouss replied:
Hermetic python is not a flag you can switch off unfortunately. Patch 0011 ensures we are using our python interpreter (which means whatever package installed in this environment is available as well). 
The tf build system does not know how about this however. Tensorflow uses https://github.com/bazelbuild/rules_python targets to configure pypi dependencies at build time (for example, `@pypi_keras//:pkg`  essentially triggers a `pip install`). A proper way would be to enhance that project to handle system dependencies instead. This is a big undertaking. 
The python dependencies are not vendored though (you can verify this by extracting the resulting tensorflow package). So at runtime, our dependencies are used as expected. 

The bazel files in the feedstock do not replace upstream files. Rather, the extra configuration is added through `.bazelrc` set in our build scripts.

danpetry replied:
> at runtime, our dependencies are used as expected

and, indeed at build time?
Thanks sounds good to me.

danpetry reviewed (commented):
I reviewed the hermetic python stuff sort of (to the best of my understanding) and still need to give a final once-over to make sure I've looked at everything.

danpetry committed:
Use patchelf support from bazel-toolchain instead

Patchelf support was transferred to bazel-toolchain and
is used from there rather than maintaining it here.

danpetry committed:
restrict python as per upstream testing

danpetry committed:
Update README's update instructions

danpetry reviewed (approved):
I've finished my review and am good with the changes once the builds pass

danpetry commented:
We also had sporadic xcode-related failures on osx-64 and osx-arm64:
```
10/02/24 10:38:16  INFO   Analyzing: 3 targets (646 packages loaded, 25696 targets configured)                                                                  
                   INFO   ERROR: /private/var/tmp/_bazel_builder/3d27507a171c46891a410d16c736a180/external/local_config_apple_cc/BUILD:61:23: in                
                          apple_cc_toolchain rule @local_config_apple_cc//:cc-compiler-darwin_arm64: Xcode version must be specified to use an Apple CROSSTOOL. 
                          If your Xcode version has changed recently, verify that "xcode-select -p" is correct and then try: "bazel shutdown" to re-run Xcode   
                          configuration                                                                                                                         
                   INFO   ERROR: /private/var/tmp/_bazel_builder/3d27507a171c46891a410d16c736a180/external/local_config_apple_cc/BUILD:61:23: Analysis of target
                          '@local_config_apple_cc//:cc-compiler-darwin_arm64' failed                                                                            
                   INFO   ERROR: /private/var/tmp/_bazel_builder/3d27507a171c46891a410d16c736a180/external/bazel_tools/src/tools/launcher/BUILD:9:14: errors    
                          encountered resolving toolchains for @bazel_tools//src/tools/launcher:launcher                                                        
                   INFO   ERROR:                                                                                                                                
                          /private/var/folders/k1/30mswbxs7r1g6zwn8y4fyt500000gp/T/abs_fdl1h7az42/croot/tensorflow-split_1727883363694/work/tensorflow/tools/bu…
                          errors encountered resolving toolchains for //tensorflow/tools/build_info:gen_build_info
```
however restarting the build solves it. It could have been differences in configuration of our (physical) osx builders. In any case, it is using our toolchain:
```
-isystem \
    /Users/dpetry/miniconda3/envs/bld/conda-bld/tensorflow-split_1728335448941/_build_env/include/c++/v1 \
    -isystem \
    /Users/dpetry/miniconda3/envs/bld/conda-bld/tensorflow-split_1728335448941/_build_env/lib/clang/14.0.6/include \
    -isystem \
    /Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk/usr/include \
    -isystem \
    /Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk/System/Library/Frameworks \
```

skupr-anaconda commented on recipe/meta.yaml:
```diff
@@ -228,7 +243,7 @@ outputs:
+        - keras >=3.2
       run:
         - python
         - packaging
```
Is `numpy` required at runtime too? See https://github.com/tensorflow/tensorflow/blob/v2.17.0/tensorflow/tools/pip_package/setup.py#L88-L89

skupr-anaconda replied:
Do we need `pyyaml` on `linux`? Or is it the Ubuntu-specific dependency? See https://github.com/tensorflow/tensorflow/blob/v2.17.0/tensorflow/tools/pip_package/setup.py#L118

skupr-anaconda replied:
It seems python 3.11 for bazel 6 requires `lit`, see https://github.com/tensorflow/tensorflow/blob/v2.17.0/tensorflow/tools/ci_build/release/requirements_common.txt#L45

skupr-anaconda replied:
Is `certifi` required on `osx`? see https://github.com/tensorflow/tensorflow/blob/v2.17.0/tensorflow/tools/ci_build/release/requirements_common.txt#L45

cbouss replied:
Yes we do need numpy. It was there through transitive dependencies but I will add it explicitly. Thanks!

pyyaml is for tests only. 

We can do without certifi and lit. What the build does is pull from, for example, https://github.com/tensorflow/tensorflow/blob/v2.17.0/requirements_lock_3_11.txt

skupr-anaconda commented on recipe/meta.yaml:
```diff
@@ -249,7 +264,7 @@ outputs:
         # TF-API needs to move in sync
         - tensorboard >=2.17,<2.18
-        - keras >=3.0
+        - keras >=3.2
```
```suggestion
        - keras >=3.2.0
```
https://github.com/tensorflow/tensorflow/blob/v2.17.0/tensorflow/tools/pip_package/setup.py#L118

cbouss committed:
remove temp channel

cbouss committed:
address comments

cbouss committed:
restrict osx to big-sur

danpetry commented:
🎉
